### PR TITLE
AWS SDKv2 Followup

### LINKED ISSUE #314: Rearchitect downloader design
**Detailed Description**
The downloader component is inefficient and needs to be rewritten to:

- Maximize the download throughput from the cloud (signatures _and_ record/balance files)
  - Use some sort of priority queue to schedule downloads
- Minimize tasks done serially in the downloader
  - currently AWS SDK v1 performs several serial HTTP requests prior to queueing the actual file download in the TransferManager
  - non-signature files are currently only downloaded after enough signatures have been downloaded for that record/event/balance stream (the actual download can be done in parallel without actually processing it or adding it to the DB until enough signatures have been verified)
- Store the files and process them in memory as opposed to on the filesystem (eliminating the use of the file system as a signal between the downloader and record processor).
- Potentially download from several cloud sources in parallel if that helps (google cloud and AWS S3 in parallel) - if downloading from a single source becomes a bottleneck
- Separate the logic that determines what files are available in the cloud (currently via polling the bucket) from the actual priority queue/download piece. We may have better alternatives to constantly polling the bucket (receiving notifications from the bucket instead of polling).

We should also consider how we want to handle "catch up" processing for standing up a clean/new node with an empty database but a long history of files in the cloud. Possibly involving prioritization of the downloads (always try to download the most recent files when available, and catch up with older files at a lower priority) - this requires more discussion.

**Actual Behavior**
- Currently using AWS sdk v1 TransferManager.
- record and account balance files (not including signatures) are serially downloaded.

**Expected Behavior**

**Environment:**

**Additional Context**
AWS SDK v2 has many improvements over v1. Downloading can be done on a single thread/event loop - eliminating the requirement for an additional thread per download (everything is IO-bound here). This is an option.

### PR DESCRIPTION
**Pull request by apeksharma**
Signed-off-by: Apekshit Sharma <apekshit.sharma@hedera.com>

**Which issue(s) this PR fixes**:
Addresses #314 partially.

**Special notes for your reviewer**:
Followup of PR #347.

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

**Comment by apeksharma** [comment]
Ignore test failures. Like all other PRs, failing tests are from two flaky ones in RecordFileLoggerContractTest.

File: src/main/java/com/hedera/mirror/downloader/Downloader.java [hunk 1]
```diff
@@ -128,18 +130,27 @@ protected void downloadNextBatch() {
...
 				// Get a list of objects in the bucket, 100 at a time
-				String prefix = s3Prefix + nodeAccountId + "/";
+				String s3prefix = prefix + nodeAccountId + "/";
 				Stopwatch stopwatch = Stopwatch.createStarted();
 
+                // prefix is of format "X/Y" (e.g. "recordstreams/record"), so a replace here with use of Paths in rest
+                // of the code ensures platform compatibility. More involved way would splitting 'prefix' in all
+                // DownloaderProperties implementations to two values and then join then separately for S3 and for
+                // local filesystem.
+                final Path sigFilesDir = downloaderProperties.getStreamPath().getParent()
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/mirror/downloader/Downloader.java]: Would be clearer to move `Path dataPath = downloaderProperties.getStreamPath().getParent();` outside _all_ loops.

File: src/main/java/com/hedera/mirror/config/MirrorNodeConfiguration.java [hunk 2]
```diff
@@ -94,16 +94,15 @@ public void customize(FluentConfiguration configuration) {
...
         AwsCredentialsProvider awsCredentials = AnonymousCredentialsProvider.create();
 
-        if (StringUtils.isNotBlank(downloaderProperties.getAccessKey()) &&
-                StringUtils.isNotBlank(downloaderProperties.getSecretKey())) {
-            awsCredentials = StaticCredentialsProvider.create(
-                    AwsBasicCredentials.create(downloaderProperties.getAccessKey(), downloaderProperties.getSecretKey()));
+        String accessKey = downloaderProperties.getAwsAccessKey();
+        String secretKey = downloaderProperties.getAwsSecretKey();
+        if (StringUtils.isNotBlank(accessKey) && StringUtils.isNotBlank(secretKey)) {
+            awsCredentials = StaticCredentialsProvider.create(AwsBasicCredentials.create(accessKey, secretKey));
```
Comment (Nana-EC) [inline src/main/java/com/hedera/mirror/config/MirrorNodeConfiguration.java]: nit: SHouldn't there be a log for the case where the keys are blank?
Maybe use the if() to catch that and log that. Then outside the if() do the awsCredentials assignment.

File: src/test/java/com/hedera/mirror/downloader/AbstractDownloaderTest.java [hunk 3]
```diff
@@ -131,6 +133,24 @@ protected void assertValidFiles(List<String> filenames) throws Exception {
                 .containsAll(filenames);
     }
 
+    protected void overwriteOnDownloadHelper(String fileName1, String fileName2, ApplicationStatusCode key)
+            throws Exception {
+        fileCopier.copy();
+        downloader.download();
+        verify(applicationStatusRepository).updateStatusValue(key, fileName1);
+        verify(applicationStatusRepository).updateStatusValue(key, fileName2);
+        assertValidFiles(List.of(fileName1, fileName2));
```
Comment (steven-sheehy) [inline src/test/java/com/hedera/mirror/downloader/AbstractDownloaderTest.java]: This test is strange and doesn't properly test the scenario. The scenario is a corrupted/incomplete signature file, not downloading the same valid file twice. The difference is if the test is incorrect the corrupt signature file will fail while this one will still pass.

Please write the corrupt signature file first to say `recordstreams/record0.0.3/` then call `downloader.download()` _once_ and confirm it continues successfully. Use code coverage to verify inside the `if (file.exists())` branch is reached.

File: src/main/java/com/hedera/mirror/downloader/Downloader.java [hunk 4]
```diff
@@ -118,38 +117,50 @@ protected void downloadNextBatch() {
...
 		 */
+        final Path dataPath = downloaderProperties.getStreamPath().getParent();
 		for (String nodeAccountId : nodeAccountIds) {
 			tasks.add(Executors.callable(() -> {
 				log.debug("Downloading signature files for node {} created after file {}", nodeAccountId, lastValidSigFileName);
+                Stopwatch stopwatch = Stopwatch.createStarted();
 				// Get a list of objects in the bucket, 100 at a time
-				String prefix = s3Prefix + nodeAccountId + "/";
-				Stopwatch stopwatch = Stopwatch.createStarted();
+				String s3prefix = downloaderProperties.getPrefix() + nodeAccountId + "/";
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/mirror/downloader/Downloader.java]: `s3Prefix` or just `prefix`

File: src/test/java/com/hedera/mirror/downloader/AbstractDownloaderTest.java (hunk 3 above)
Comment (apeksharma) [reply src/test/java/com/hedera/mirror/downloader/AbstractDownloaderTest.java]: i like the point of corrupting them to be extra sure.
I verified that `file.delete()` was being called by putting breakpoint. Also verified in code coverage report now.

File: src/main/java/com/hedera/mirror/downloader/Downloader.java (hunk 4 above)
Comment (apeksharma) [reply src/main/java/com/hedera/mirror/downloader/Downloader.java]: Done.

**Pull request merged**